File: api/mpesa_callback.py
from flask import request, jsonify
from datetime import datetime
import pymysql
import json

# ================================================================
# DATABASE CONFIGURATION
# ================================================================
DB_CONFIG = {
    'host': 'localhost',
    'user': 'root',
    'passwd': '',
    'db': 'zilla',
    'charset': 'utf8mb4'
}

def get_db():
    """Get database connection"""
    try:
        conn = pymysql.connect(
            host=DB_CONFIG['host'],
            user=DB_CONFIG['user'],
            password=DB_CONFIG['passwd'],
            database=DB_CONFIG['db'],
            charset=DB_CONFIG['charset']
        )
        return conn
    except pymysql.Error as e:
        print(f"Database connection error: {e}")
        return None

def handle_mpesa_callback(app):
    """Register M-PESA callback endpoint"""
    
    @app.route('/api/mpesa_callback', methods=['POST'])
    def mpesa_callback():
        print("\n" + "=" * 50)
        print("📲 M-PESA CALLBACK RECEIVED")
        print("=" * 50)
        
        try:
            # Get callback data
            callback_data = request.json
            print(f"Callback data: {json.dumps(callback_data, indent=2)}")
            
            # Extract the STK callback
            stk_callback = callback_data.get('Body', {}).get('stkCallback', {})
            
            merchant_request_id = stk_callback.get('MerchantRequestID')
            checkout_request_id = stk_callback.get('CheckoutRequestID')
            result_code = stk_callback.get('ResultCode')
            result_desc = stk_callback.get('ResultDesc')
            
            print(f"   MerchantRequestID: {merchant_request_id}")
            print(f"   CheckoutRequestID: {checkout_request_id}")
            print(f"   ResultCode: {result_code}")
            print(f"   ResultDesc: {result_desc}")
            
            conn = get_db()
            if not conn:
                print("❌ Database connection failed")
                return jsonify({'ResultCode': 0, 'ResultDesc': 'Accepted'})
            
            cursor = conn.cursor(pymysql.cursors.DictCursor)
            
            try:
                # Find the transaction
                cursor.execute("""
                    SELECT * FROM mpesa_transactions 
                    WHERE checkout_request_id = %s
                """, (checkout_request_id,))
                
                transaction = cursor.fetchone()
                
                if not transaction:
                    print(f"❌ Transaction not found for checkout_request_id: {checkout_request_id}")
                    return jsonify({'ResultCode': 0, 'ResultDesc': 'Accepted'})
                
                sale_id = transaction['sale_id']
                
                if result_code == 0:
                    # Payment successful
                    print("✓ Payment successful!")
                    
                    # Extract callback metadata
                    callback_metadata = stk_callback.get('CallbackMetadata', {}).get('Item', [])
                    
                    mpesa_receipt_number = None
                    transaction_date = None
                    phone_number = None
                    amount = None
                    
                    for item in callback_metadata:
                        name = item.get('Name')
                        value = item.get('Value')
                        
                        if name == 'MpesaReceiptNumber':
                            mpesa_receipt_number = value
                        elif name == 'TransactionDate':
                            transaction_date = value
                        elif name == 'PhoneNumber':
                            phone_number = value
                        elif name == 'Amount':
                            amount = value
                    
                    print(f"   M-PESA Receipt: {mpesa_receipt_number}")
                    print(f"   Amount: {amount}")
                    print(f"   Phone: {phone_number}")
                    
                    # Update transaction
                    cursor.execute("""
                        UPDATE mpesa_transactions 
                        SET status = 'completed',
                            mpesa_receipt_number = %s,
                            result_code = %s,
                            result_desc = %s,
                            updated_at = NOW()
                        WHERE id = %s
                    """, (mpesa_receipt_number, result_code, result_desc, transaction['id']))
                    
                    # Update sale
                    cursor.execute("""
                        UPDATE sales 
                        SET transaction_status = 'completed',
                            paid_amount = %s,
                            mpesa_receipt_number = %s,
                            updated_at = NOW()
                        WHERE id = %s
                    """, (amount, mpesa_receipt_number, sale_id))
                    
                    conn.commit()
                    print(f"✓ Transaction and sale updated successfully")
                    
                else:
                    # Payment failed
                    print(f"❌ Payment failed: {result_desc}")
                    
                    # Update transaction
                    cursor.execute("""
                        UPDATE mpesa_transactions 
                        SET status = 'failed',
                            result_code = %s,
                            result_desc = %s,
                            updated_at = NOW()
                        WHERE id = %s
                    """, (result_code, result_desc, transaction['id']))
                    
                    # Update sale
                    cursor.execute("""
                        UPDATE sales 
                        SET transaction_status = 'failed',
                            updated_at = NOW()
                        WHERE id = %s
                    """, (sale_id,))
                    
                    conn.commit()
                    print(f"✓ Transaction marked as failed")
                
            except pymysql.Error as e:
                print(f"❌ Database error: {e}")
                conn.rollback()
            finally:
                cursor.close()
                conn.close()
            
            # Always return success to M-PESA
            return jsonify({'ResultCode': 0, 'ResultDesc': 'Accepted'})
            
        except Exception as e:
            print(f"❌ Callback error: {str(e)}")
            import traceback
            traceback.print_exc()
            return jsonify({'ResultCode': 0, 'ResultDesc': 'Accepted'})

File: s/api/mpesa_callback.php
<?php
/**
 * ================================================================
 * M-PESA PAYMENT CALLBACK HANDLER
 * ================================================================
 * This endpoint receives callbacks from M-Pesa when payments are made
 * Handler for both: STK Push & Query responses
 * 
 * URL: http://13.57.193.106/api/mpesa_callback.php
 * ================================================================
 */

require_once dirname(__DIR__) . '/config.php';

// Allow callback from M-Pesa without authentication
header('Content-Type: application/json');

// Log all incoming requests
$input = file_get_contents("php://input");
error_log("M-Pesa Callback Received: " . $input);

// Decode JSON payload
$data = json_decode($input, true);

if (!$data) {
    http_response_code(200); // Always return 200 to M-Pesa
    echo json_encode(['success' => false, 'message' => 'Invalid JSON payload']);
    exit;
}

// Verify it's a valid callback (basic security)
if (empty($data['Body'])) {
    http_response_code(200);
    echo json_encode(['success' => false, 'message' => 'Invalid callback structure']);
    exit;
}
}

// Extract the actual result
$body = $data['Body']['stkCallback'] ?? $data['Body'];
$resultCode = $body['ResultCode'] ?? null;
$resultDesc = $body['ResultDesc'] ?? 'Unknown error';

// Extract checkout request ID and merchant request ID
$checkoutRequestID = $body['CheckoutRequestID'] ?? null;
$merchantRequestID = $body['MerchantRequestID'] ?? null;

// Log callback details
error_log("M-Pesa Callback - Result Code: {$resultCode}, Description: {$resultDesc}");

// ================================================================
// HANDLE DIFFERENT RESPONSE CODES
// ================================================================

switch ($resultCode) {
    
    // ============ SUCCESS (0) ============
    case '0':
    case 0:
        handlePaymentSuccess($body);
        break;
    
    // ============ CANCELLED (1032) ============
    case '1032':
    case 1032:
        error_log("M-Pesa Transaction Cancelled: {$checkoutRequestID}");
        // Update transaction status in database if needed
        updateTransactionStatus($checkoutRequestID, 'cancelled');
        break;
    
    // ============ TIMEOUT (1037) ============
    case '1037':
    case 1037:
        error_log("M-Pesa Transaction Timeout: {$checkoutRequestID}");
        updateTransactionStatus($checkoutRequestID, 'timeout');
        break;
    
    // ============ INSUFFICIENT FUNDS (1001) ============
    case '1001':
    case 1001:
        error_log("M-Pesa Insufficient Funds: {$checkoutRequestID}");
        updateTransactionStatus($checkoutRequestID, 'insufficient_funds');
        break;
    
    // ============ OTHER FAILURES ============
    default:
        error_log("M-Pesa Payment Failed ({$resultCode}): {$checkoutRequestID} - {$resultDesc}");
        updateTransactionStatus($checkoutRequestID, 'failed', $resultDesc);
}

// Return success response to M-Pesa (must always return 200 OK)
http_response_code(200);
echo json_encode(['success' => true, 'message' => 'Callback received']);
exit;

// ================================================================
// HELPER FUNCTIONS
// ================================================================

/**
 * Handle successful payment
 */
function handlePaymentSuccess($body) {
    $checkoutRequestID = $body['CheckoutRequestID'] ?? null;
    $merchantRequestID = $body['MerchantRequestID'] ?? null;
    
    // Extract payment details from CallbackMetadata
    $callbackMetadata = $body['CallbackMetadata']['Item'] ?? [];
    $mpesaReceiptNumber = '';
    $transactionDate = '';
    $phoneNumber = '';
    $amount = '';
    
    // Parse callback metadata
    foreach ($callbackMetadata as $item) {
        $name = $item['Name'] ?? '';
        $value = $item['Value'] ?? '';
        
        switch ($name) {
            case 'MpesaReceiptNumber':
                $mpesaReceiptNumber = $value;
                break;
            case 'TransactionDate':
                $transactionDate = $value;
                break;
            case 'PhoneNumber':
                $phoneNumber = $value;
                break;
            case 'Amount':
                $amount = $value;
                break;
        }
    }
    
    error_log("M-Pesa Payment Success - Receipt: {$mpesaReceiptNumber}, Amount: {$amount}, Phone: {$phoneNumber}");
    
    try {
        // Update the mpesa_transactions table with payment confirmation
        $stmt = db()->prepare("
            UPDATE mpesa_transactions 
            SET status = ?, 
                mpesa_receipt = ?,
                transaction_date = NOW(),
                result_code = '0',
                result_desc = 'Success'
            WHERE checkout_request_id = ?
        ");
        $stmt->execute(['completed', $mpesaReceiptNumber, $checkoutRequestID]);
        
        error_log("M-Pesa Transaction confirmed: {$checkoutRequestID} - Receipt: {$mpesaReceiptNumber}");
        
    } catch (Exception $e) {
        error_log("Error processing M-Pesa callback: " . $e->getMessage());
    }
}

/**
 * Update transaction status
 */
function updateTransactionStatus($checkoutRequestID, $status, $reason = '') {
    try {
        $stmt = db()->prepare("
            UPDATE mpesa_payments 
            SET status = ?, reason = ?, updated_at = NOW()
            WHERE checkout_request_id = ?
        ");
        $stmt->execute([$status, $reason, $checkoutRequestID]);
        
        error_log("Transaction status updated: {$checkoutRequestID} -> {$status}");
        
    } catch (Exception $e) {
        error_log("Error updating transaction status: " . $e->getMessage());
    }
}
